Switches project setup from Maven to Gradle an provides shared configs as Gradle plugin

cschierle opened the pull request:
Fixes: SIRI-632


### Breaking changes:
- Maven setup (e.g. `pom.xml`) has been replaced by Gradle equivalent
- The [Gradle plugin](https://plugins.gradle.org/plugin/com.scireum.sirius-parent) in `SiriusParentPlugin.kt` now provides the shared config for sirius projects
- Drone config in `drone.yml` has been updated to use Gradle tasks for validation and publication

sabieber committed:
Migrates Maven parent functionality to Gradle plugin

Fixes: OX-5764

sabieber committed:
Adds copy kotlin marker task to builds

Fixes: MISC-3

sabieber committed:
Adds test dependencies to projects using the plugin

Fixes: MISC-3

sabieber committed:
Configures kotlin in projects using the plugin

Fixes: MISC-3

sabieber committed:
Adds plugin publishing information

sabieber committed:
Executes all tests on the junit platform

This is necessary as JUnit 5 does not support test suites and we want to run both old junit 4 and new junit 5 tests.

sabieber committed:
Allows to configure the source uri of intellij settings

sabieber committed:
Adjusts plugin id to match specifications

sabieber committed:
Sets kotlin compiler target to Java 16

Fixes: OX-7895

sabieber committed:
Upgrades dependencies used for testing

Fixes: OX-7895

sabieber committed:
Merge branch 'master' into sbi/NewGradlePlugin

# Conflicts:
#	pom.xml

sabieber committed:
Upgrades kotlin compiler target to Java 17

sabieber committed:
Upgrades testing dependencies

sabieber committed:
Upgrades project language level to Java 17

sabieber committed:
Merge branch 'master' into sbi/NewGradlePlugin

# Conflicts:
#	pom.xml

sabieber committed:
Upgrades gradle wrapper to version 7.3.3

Fixes: MISC-51

sabieber committed:
Upgrades kotlin libraries to version 1.6.10

Fixes: MISC-51

sabieber committed:
Switches jUnit test dependencies to correct configuration scope

The testRuntime scope does not seem to be supported anymore.

Fixes: MISC-51

sabieber committed:
Removes jCenter maven repo as it was killed

Fixes: MISC-51

sabieber committed:
Merge branch 'master' into sbi/NewGradlePlugin

# Conflicts:
#	pom.xml

sabieber committed:
Includes jUnit lib for parameterized tests

Fixes: MISC-51

sabieber committed:
Includes Kotlin and MockK for new tests

We want to replace our old Groovy + Spock + jUnit 4 Setup with Kotlin + MockK + jUnit 5 for our automated tests.
This includes the necessary dependencies and build plugins to also run the new tests (next to old ones).

Fixes: MISC-51

sabieber committed:
Upgrades kotlin version to 1.6.21

sabieber committed:
Upgrades java target version to 18

sabieber committed:
Merge branch 'master' into sbi/NewGradlePlugin

sabieber committed:
Excludes gradle files from git via ignore file

Fixes: SIRI-577

sabieber committed:
Updates plugin version to RC

Fixes: SIRI-577

sabieber committed:
Updates supplied kotlin version to 1.7.10

Fixes: SIRI-577

sabieber committed:
Updates gradle publish plugin to version 1.0.0

Fixes: SIRI-577

sabieber committed:
Provides documentation of plugin publish process

Fixes: SIRI-577

sabieber committed:
Extends documentation

Fixes: SIRI-577

sabieber committed:
Excludes IntelliJ files from git via ignore file

The actual IntelliJ files of the plugin project itself should not be pushed to git. The ones rolled out to projects using the plugins are in the resources directory.

Fixes: SIRI-577

Commit:
Prevents duplicate test entities in the build directory when running tests

Fixes: SIRI-576

Commit:
Upgrades to Gradle 7.5

Fixes: SIRI-576

Commit:
Switches to the new way of setting source directories.
The previous way is deprecated and marked for removal with Gradle 8.

Fixes: SIRI-576

Commit:
Increases rc number for a new release

Fixes: SIRI-576

Commit:
Sets duplicates strategy for jar task

Fixes: SIRI-578

Commit:
Optimizes test dependencies

Fixes: SIRI-632

Commit:
Adds custom test task without nightly tests

Fixes: SIRI-632

Commit:
Extracts functions for better readability

Fixes: SIRI-632

Commit:
Outputs sirius logs and test failures when running tests

Fixes: SIRI-632

Commit:
Changes scope of jupiter engine dependency to test runtime

Fixes: SIRI-632

Commit:
Fixes typo

Fixes: SIRI-632

Commit:
Marks version as new release candidate

Fixes: SIRI-632

Commit:
Moves testWithoutNightly to verification category

Fixes: SIRI-634

Commit:
Renames CopyMarkerTask to CopyMarkerAction

Fixes: SIRI-634

Commit:
Uses a simple action to copy component markers.
This eliminates complex implicit task dependencies, which will no longer be automatically resolved when upgrading to Gradle 8.

Fixes: SIRI-634

Commit:
Moves test task setup into TaskContainer

Fixes: SIRI-634

Commit:
Sets up Jacoco reporting to use with SonarQube

Fixes: SIRI-634

Commit:
Merge pull request #31 from scireum/feature/csc/GradleCleanup

Cleanup and Jacoco setup

sabieber commented on src/main/kotlin/sirius/parent/SiriusParentPlugin.kt:
```diff
@@ -107,32 +107,28 @@ class SiriusParentPlugin : Plugin<Project> {
-            testTaskWithoutNightly.useJUnitPlatform { platformOptions ->
-                platformOptions.excludeTags = setOf("nightly")
-            }
+    private fun TaskContainer.initializeTestTaskWithoutNightly() {
```
Was macht ein TaskContainer?

cschierle replied:
Der sagt einfach, dass die Methode innerhalb von so einem `project.tasks.apply { ... }` Block aufrufbar ist. Dachte das macht mehr Sinn für die Methoden wo's um Task setup geht.

Gibt theoretisch auch `DependencyContainer`, `PluginContainer`, usw. Aber dort haben wir aktuell nicht genug Setup um das in mehrere Methoden aufzuteilen

Commit:
Sets up Maven Publish plugin

Fixes: SIRI-633

Commit:
Allows customizable uri and credentials

Fixes: SIRI-633

Commit:
Reads credentials directly from system env.
Required for passing secrets in drone builds.

Fixes: SIRI-633

Commit:
Copies component markers after testClasses task

Fixes: SIRI-633

Commit:
Merge branch 'sbi/NewGradlePlugin' into feature/csc/SIRI-633_publish

Commit:
Merge pull request #32 from scireum/feature/csc/SIRI-633_publish

Sets up maven publish plugin

Commit:
Bumps rc number

Commit:
Ports drone stages to gradle

Fixes: SIRI-723

Commit:
Also publishes plugin to scireum-mvn

Fixes: SIRI-723

Commit:
Bumps plugin versions

Fixes: SIRI-723

Commit:
Uses platform BOM dependencies to simplify version alignment of included libraries

Fixes: SIRI-723

Commit:
Upgrades dependencies

Fixes: SIRI-723

Commit:
Merge branch 'master' into sbi/NewGradlePlugin

# Conflicts:
#	pom.xml

Commit:
Upgrades to Gradle 8.1.1 and simplifies kotlin dependency setup

Fixes: SIRI-723

Commit:
Extracts method for test task setup

Fixes: SIRI-723

Commit:
Adds test task without class filter.
This is to be used when executing individual test cases locally.

Fixes: SIRI-723

Commit:
Properly configures Kotlin test source set.
This is needed if a test references test entities.

Fixes: SIRI-723

Commit:
Properly configures test inclusions.
`test` and `testWithoutNightly` run Groovy Specs via "TestSuite" configurations under JUnit 4, while Java and Kotlin tests (named `...Test.(kt|java)`) are run under JUnit 5 with "SiriusExtension"s where applicable

Fixes: SIRI-723

Commit:
Downgrades to Gradle version 7.6.1.
Kotlin language version 1.8.21 does not yet officially support Gradle 8.

Fixes: SIRI-723

Commit:
Properly names the scireum-mvn repo.
Simplifies identifying if a dependency was fetched from there or from general maven repository.

Fixes: SIRI-723

sabieber committed:
Provides jackson bom to all projects via plugin

Commit:
Adds component marker to tests jar by depending on testClasses

Fixes: SIRI-723

Commit:
Simplifies variable assignment

Fixes: SIRI-723

Commit:
Applies application plugin to projects using this plugin.
This provides the proper configuration for the Gradle 'run' task.

Fixes: SIRI-723

Commit:
Simplifies task dependency setup

Fixes: SIRI-723

mkeckmkeck reviewed (approved):
lgtm, kleiner walkthrough durch SiriusParentPlugin.kt bei gelegenheit wär cool

Commit:
Clarifies method purpose

Fixes: SIRI-818

Commit:
Removes unnecessary test task.
We intend to run tests locally directly through IntelliJ IDE, so we don't need this extra task.

Fixes: SIRI-818

Commit:
Adds component.marker to main classes for proper classpath building when executing projects locally

Fixes: SIRI-818

Commit:
Slightly improves test setup

Fixes: SIRI-818

Commit:
Provides run config for local install

Fixes: SIRI-818
